File: SearchInsertPosition.py
"""
35. Search Insert Position
Given a sorted array and a target value, return the index if the target is found. If not, return the index where it would be if it were inserted in order.
You may assume no duplicates in the array.
Here are few examples.
[1,3,5,6], 5 → 2
[1,3,5,6], 2 → 1
[1,3,5,6], 7 → 4
[1,3,5,6], 0 → 0
"""
# O(logn)
# O(1)
class Solution(object):
    def searchInsert(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: int
        """
        lower, upper = 0, len(nums) - 1
        while lower < upper:
            mid = (lower + upper) // 2
            if nums[mid] == target:
                return mid
            elif nums[mid] > target:
                upper = mid - 1
            else:
                lower = mid + 1
        if nums[lower] >= target:
            return lower
        return lower + 1

if __name__ == "__main__":
    print(Solution().searchInsert([1,3,5,6], 2))

File: Java/SearchInsertPosition.java

public class SearchInsertPosition {
    public static int searchInsert(int[] nums, int target) {
        int lower = 0;
        int upper = nums.length - 1;
        int mid;
        while(lower <= upper) {
            mid = (lower + upper) / 2;
            if(nums[mid] == target) {return mid;}
            if(nums[mid] < target) {
                lower = mid + 1;
            } else {
                upper = mid - 1;
            }
        }
        return lower;
    }

	public static void main(String[] args) {
		int[] nums = {1, 3, 5, 6};
		System.out.println(searchInsert(nums, 5));
		System.out.println(searchInsert(nums, 2));
		System.out.println(searchInsert(nums, 7));
		System.out.println(searchInsert(nums, 0));
	}

}
